File: R/booknet.R
#' BookNet
#'
#' Reads a directory of text files and generates
#' an adjacency matrix of characters in those text files, with weights
#' equal to the number of times the relationship appeared.
#' A connection is made when characters appear within
#' window_length tokens of each other, delimited by spaces.
#' The directory must contain one "names.txt" file which
#' consists of Java regular expressions one each line to search for, for a
#' given character, ie "Jaime|Kingslayer" would match for
#' Jaime Lannister in ASOIAF.
#'
#' @param folder_path path to a directory which contains the *.txt files (including "names.txt")
#' @param vertex_names optional - names for each row/column in the matrix, and vertex labels for `igraph`.
#' Must be specified in the same order as in names.txt. If missing labels from names.txt will be used instead.
#' @param window_length window of words to check for character occurences.
#' @param min_occurences weights in the matrix below this number will be set to zero.
#' @param percent_remove removes the bottom `percent_remove` percentage of nonzero weights.
#' If `min_occurences` is specified, those occurences will be removed first, before removing
#' the bottom percentage specified by `percent_remove`.
#'
#' @return a list, with `$adjacency_matrix` returning the labeled adjacency matrix,
#' and `$adj_graph` returning an `igraph` graph.
#' @export
#'
#' @examples
#' out <- booknet("data/ASOIAF_Folder", min_occurences = 3)
booknet <- function(folder_path, vertex_names, window_length = 15, min_occurences = 0, percent_remove = 0) {
  class_name <- J("BookNet")
  full_path <- gsub("\\\\", "//",normalizePath(folder_path))
  book_net <- new(class_name, full_path, as.integer(window_length))
  .jcall(book_net, "V", "doAll")
  out_matrix <- .jcall(book_net, "[[D", "getAdjacencyMatrix")
  to_return <- matrix(nrow = 0, ncol = length(out_matrix))
  for (row in out_matrix) {
    to_return <- rbind(to_return, .jevalArray(row))
  }

  if (missing(vertex_names)) {
    vertex_names <- .jcall(book_net, "[S", "getNames")
  }
  if (length(vertex_names) != ncol(to_return)) {
    stop("vertex_names is not the correct length!")
  }

  colnames(to_return) <- vertex_names
  rownames(to_return) <- vertex_names

  if (min_occurences != 0) {
    for (r in 1:nrow(to_return)) {
      for (c in 1:ncol(to_return)) {
        if (to_return[r,c][[1]] < min_occurences) {to_return[r,c] <- 0}
      }
    }
  }

  if (percent_remove != 0) {
    if (percent_remove < 0 | percent_remove > 1) stop("min_cap must be between 0 and 1")
    min_cap <- quantile(to_return[which(to_return > 0)], percent_remove)[[1]]
    for (r in 1:nrow(to_return)) {
      for (c in 1:ncol(to_return)) {
        if (to_return[r,c][[1]] < min_cap) {to_return[r,c] <- 0}
      }
    }
  }

  adj_graph <- graph_from_adjacency_matrix(to_return, mode="undirected", weighted = TRUE, add.colnames = 'label')
  return(list(adjacency_matrix = to_return, graph = adj_graph))
}

File: java/src/BookNet.java
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.util.*;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Object that reads a directory of text files and generates
 * an adjacency matrix of characters in those text files.
 * A connection is made when characters appear within
 * queueSize tokens of each other, delimited by spaces.
 * The directory must contain one "names.txt" file which
 * consists of regular expressions one each line to search for, for a
 * given character, ie "Jaime|Kingslayer" would match for
 * Jaime Lannister in ASOIAF.
 */
public class BookNet {
    private int queueSize;
    private String directory;
    private AdjacencyMatrix matrix;
    private List<String> names;

    /**
     * Initializes the BookNet object
     * @param directory path of the directory containing names.txt and other txt files
     * @param queueSize size of the window to check for matches
     */
    public BookNet(String directory, int queueSize) {
        this.directory = directory;
        names = new ArrayList<>();
        readNames();
        matrix = new AdjacencyMatrix(names);
        this.queueSize = queueSize;
    }

    /**
     * Get the regex of characters
     * @return string array of regular expressions of characters
     */
    public String[] getNames() {
        return names.toArray(new String[names.size()]);
    }

    /**
     * Read in the names from the names.txt file
     * @throws NoSuchFileException when names.txt not found
     */
    private void readNames() {
        FileInputStream nameStream;

        try {
            nameStream = new FileInputStream(directory + "//names.txt");
        } catch (FileNotFoundException e) {
            throw new NoSuchFileException();
        }
        Scanner scanner = new Scanner(nameStream);
        while (scanner.hasNext()) {
            String next = scanner.nextLine();
            names.add(next);
        }
    }

    /**
     * Calles doOneFile for every text file, on a new thread
     */
    public void doAll() {
        File[] fileArray = new File(directory).listFiles();
        List<File> fileList = Arrays.asList(fileArray);
        fileList = new ArrayList<>(fileList);
        fileList.remove(new File(directory + "//names.txt"));
        fileList
                .stream()
                .parallel()
                .map(File::getAbsolutePath)
                .forEach(this::doOneFile);
    }

    /**
     * Get the adjacency matrix
     * @return a double[][] adjacency matrix, for R
     */
    public double[][] getAdjacencyMatrix() {
        return matrix.getAdjacencyMatrix();
    }

    /**
     * For a given file, will iterate along on a queueSize-token
     * window. On each iteration, will put every character regular
     * expression that it sees in the window into a set. Every possible
     * pairing of these characters will be generated, and each pairing
     * will be incremented by one in the adjacency matrix.
     *
     * Additionally, a map of how many tokens a given token pair was seen
     * is kept. If a pair is seen, the count is set to queueSize. Then, at
     * each iteration, every count is decremented by one. This ensures that
     * if a pair is "seen," it cannot be seen again for queueSize tokens.
     * @param fullFilePath path of the file to analyze
     */
    private void doOneFile(String fullFilePath) {
        List<Matcher> matchers = new ArrayList<>(names.size());
        for (String name:names) {
            matchers.add(Pattern.compile(name).matcher(""));
        }
        //Create a scanner for the file
        FileInputStream stream;
        try {
            stream = new FileInputStream(fullFilePath);
        } catch (FileNotFoundException e) {
            throw new NoSuchFileException();
        }
        Scanner scanner = new Scanner(stream);

        //initialize a map of character pairs that have been seen
        Map<Pair, Count> seen = new HashMap<>();
        for (Pair pair:Pair.pairUp(names)) {
            seen.put(pair, new Count(0));
        }
        //Initialize the queue
        Queue<String> words = new LinkedBlockingQueue<>(queueSize);

        while (scanner.hasNext()) {
            //"Age" every entry in the seen map
            for (Map.Entry<Pair, Count> entry:seen.entrySet()) {
                entry.getValue().decrement();
            }

            //Add the new word to the queue, removing the oldest
            String token = scanner.next();
            boolean offered = words.offer(token);
            if (!offered) {
                words.remove();
                words.add(token);
            }

            //Build a string from the queue to apply regex
            StringBuilder currentWindowBuilder = new StringBuilder();
            for (String word:words) {
                currentWindowBuilder.append(" ");
                currentWindowBuilder.append(word);
            }
            String currentWindow = currentWindowBuilder.toString();

            //Initialize a set for characters in this window
            Set<String> inThisWindow = new HashSet<>();

            //Add every character I see to inThisWindow
            for (int i = 0; i < names.size(); i++) {
                matchers.get(i).reset(currentWindow);
                if (matchers.get(i).find()) {
                    inThisWindow.add(names.get(i));
                }
            }

            for (Pair pair:Pair.pairUp(inThisWindow)) {
                if (seen.get(pair).getCount() > 0) continue;
                matrix.increment(pair);
                seen.get(pair).setCount(queueSize);
            }
        }
    }

    public static void main(String[] args) {
        BookNet net = new BookNet("data", 15);
        net.doAll();
        System.out.println(net.getNames().length);
    }
}
